proof/refine/ARM/Reply_R.thy

```diff
@@ -634,4 +618,486 @@ lemma replyRemoveTCB_invs':
               simp: cteCaps_of_def)
   done
 
-end
\ No newline at end of file
+crunches get_reply
+  for inv[wp]: P
+
+(* FIXME RT: move this to lib *)
+lemma partial_inv_inj_Some:
+  "inj f \<Longrightarrow> partial_inv f (f r) = Some r"
+  by (fastforce simp: proj_inj)
+
+(* FIXME RT: move to non det lemma bucket *)
+lemma put_id_return:
+  "put s s = return () s"
+  by (clarsimp simp: put_def return_def)
+
+lemma set_reply_obj_ref_noop:
+  "monadic_rewrite False True (reply_at rptr)
+   (return ())
+   (set_reply_obj_ref (K id) rptr x)"
+  by (clarsimp simp: set_simple_ko_def monadic_rewrite_def exec_gets
+                     update_sk_obj_ref_def
+                     get_simple_ko_def partial_inv_inj_Some get_object_def bind_assoc obj_at_def
+                     is_reply_def2 set_object_def exec_get put_id_return)
+
+lemma updateReply_corres:
+  assumes
+    rr: "\<And>r1 r2. reply_relation r1 r2 \<Longrightarrow>
+           reply_relation (g (\<lambda>y. new) r1) (f (\<lambda>_. new) r2)"
+  shows
+  "corres dc (reply_at rp) (reply_at' rp
+              and obj_at' (\<lambda>ko. replyPrev_same (f (\<lambda>_. new) ko) ko) rp)
+            (set_reply_obj_ref g rp new)
+            (updateReply rp (f (\<lambda>_. new)))"
+  apply (simp add: update_sk_obj_ref_def updateReply_def)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF _ get_reply_corres])
+      apply (rule set_reply_corres)
+      apply (simp add: rr)
+  apply wpsimp+
```

Comment:
Indentation much be off here. And the formatting of the conclusion of the lemma is off